src/Microsoft.SqlTools.ServiceLayer/Cms/CmsService.cs

```diff
@@ -0,0 +1,364 @@
+﻿//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+using Microsoft.SqlServer.Management.Common;
+using Microsoft.SqlServer.Management.RegisteredServers;
+using Microsoft.SqlServer.Management.Sdk.Sfc;
+using Microsoft.SqlTools.Hosting.Protocol;
+using Microsoft.SqlTools.ServiceLayer.Cms.Contracts;
+using Microsoft.SqlTools.ServiceLayer.Connection;
+using Microsoft.SqlTools.ServiceLayer.Connection.Contracts;
+using Microsoft.SqlTools.ServiceLayer.Hosting;
+using Microsoft.SqlTools.Utility;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Microsoft.SqlTools.ServiceLayer.Cms
+{
+    /// <summary>
+    /// Main class for DacFx service
+    /// </summary>
+    public class CmsService
+    {
+        private static object lockObject = new object();
+        private static ConnectionService connectionService = null;
+        private static RegisteredServersStore registerdServerStore = null;
+        private static readonly Lazy<CmsService> instance = new Lazy<CmsService>(() => new CmsService());
+
+        /// <summary>
+        /// Gets the singleton instance object
+        /// </summary>
+        public static CmsService Instance
+        {
+            get { return instance.Value; }
+        }
+
+        #region Public methods
+
+        /// <summary>
+        /// Initializes the service instance
+        /// </summary>
+        /// <param name="serviceHost"></param>
+        public void InitializeService(ServiceHost serviceHost)
+        {
+            serviceHost.SetRequestHandler(CreateCentralManagementServerRequest.Type, this.HandleCreateCentralManagementServerRequest);
+            serviceHost.SetRequestHandler(ListRegisteredServerRequest.Type, this.HandleListRegisteredServersRequest);
+            serviceHost.SetRequestHandler(AddRegisteredServerRequest.Type, this.HandleAddRegisteredServerRequest);
+            serviceHost.SetRequestHandler(RemoveRegisteredServerRequest.Type, this.HandleRemoveRegisteredServerRequest);
+            serviceHost.SetRequestHandler(AddServerGroupRequest.Type, this.HandleAddServerGroupRequest);
+            serviceHost.SetRequestHandler(RemoveServerGroupRequest.Type, this.HandleRemoveServerGroupRequest);
+        }
+
+        public async Task HandleCreateCentralManagementServerRequest(CreateCentralManagementServerParams createCmsParams, RequestContext<ListRegisteredServersResult> requestContext)
+        {
+            Logger.Write(TraceEventType.Verbose, "HandleCreateCentralManagementServerRequest");
```

Comment:
I noticed none of the handlers are async. If there are any long running operations we should make sure they are async so we don't block the ipc handler.

Reply:
Sure will do for new connections. I noticed we do async notification for new connection but not for list databases like scenarios - so will follow the same.

Reply:
I wouldn't recommend following the connection service example. The notification system it uses is outdated and complicated.

Reply:
Just run async tasks like https://github.com/Microsoft/sqltoolsservice/blob/master/src/Microsoft.SqlTools.ServiceLayer/QueryExecution/QueryExecutionService.cs#L230